### PR TITLE
[TDD]: implement query function to satisfy tests

### PR DESCRIPTION
**Pull request by kiprot**
- implemented the query function to satisfy the provided test requirements
- added more tests to FindMeetingQueryTest.java to test for more edge cases, and made changes in the query function to take care of the edge cases
- query function passes all the 25 tests

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +281,100 @@ public void notEnoughRoom() {
...
+            Arrays.asList(OPTIONAL_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(OPTIONAL_B))
+      );
+
+      MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_1_HOUR);
+      request.addOptionalAttendee(OPTIONAL_A);
+      request.addOptionalAttendee(OPTIONAL_B);
+
+      Collection<TimeRange> actual = query.query(events, request);
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:375]: Is this test complete or not?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -212,7 +220,10 @@ public void justEnoughRoom() {
         new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
             Arrays.asList(PERSON_A)),
         new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, true),
-            Arrays.asList(PERSON_A)));
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0830AM, TIME_0845AM, false),
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Should this become a new test?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 1 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:375]: now complete. i have added the assert function, and the test passed

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 2 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: yes. i already included the test on line 338 considerNoMandatoryAttendees(), (but forgot to delete the optional attendee before pushing code)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,119 @@
...
+    }
+
+    if(events.size() > 0) {
+
+        // Adds TimeRanges for events that required attendees will be attending.
+        for (Event event : events) {
+            for(String attendee : event.getAttendees()) {
+                if(attendeesInRequest.contains(attendee)) {
+                    eventsTimeRangeList.add(event.getWhen());
+                }
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You can add timeRangeList and optionalTimeRangeList in one pass:

if(attendeesInRequest.contains(attendee)) {
  eventsTimeRangeList.add(event.getWhen());
}
if (optionalAttendees.contains(attendee)) {
  optionalsEventTimeRangeList.add(event.getWhen());
}

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,119 @@
...
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> eventsTimeRangeList = new ArrayList<>();
+    ArrayList<TimeRange> optionalsEventTimeRangeList = new ArrayList<>();
+    
+    // Returns no options if duration is longer than a day.
+    if(request.getDuration() >  TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+
+    if(events.size() > 0) {
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This check is not necessary as 0 events it is covered by:
if(eventsTimeRangeList.size() == 0 && optionalsEventTimeRangeList.size() == 0)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,119 @@
...
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendeesInRequest = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> eventsTimeRangeList = new ArrayList<>();
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: Consider using Collection<TimeRange> to avoid processing duplicated TimeRange.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,119 @@
...
+            optionalsSlots = findSlots(optionalsEventTimeRangeList, request);
+        }
+
+        // Store available TimeRange slots for required attendees.
+        ArrayList<TimeRange> mandatorySlots = new ArrayList<>();
+        if(eventsTimeRangeList.size() > 0) {
+            mandatorySlots = findSlots(eventsTimeRangeList, request);
+        }
+
+        // Finds slots that are common to both requuired and optional attendees and returns them
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: requuired -> required

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,119 @@
...
+        // if there are any.
+        ArrayList<TimeRange> combinedSlots = new ArrayList<>();
+        if(optionalsSlots.size() > 0) {
+            for(TimeRange optionalSlot : optionalsSlots) {
+                for(TimeRange mandatorySlot : mandatorySlots) {
+                    if(mandatorySlot.equals(optionalSlot) || optionalSlot.contains(mandatorySlot)) {
+                        combinedSlots.add(mandatorySlot);
+                    } else if(mandatorySlot.contains(optionalSlot)) {
+                        combinedSlots.add(optionalSlot);
+                    }
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Consider add a function in TimeRange to return overlap of two range:

if mandatorySlot.overlaps(optionalSlot) {
    combinedSlots.Add(TimeRange.fromStartEnd(max_of_start, min_of_end))
}

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,119 @@
...
+        // Store available TimeRange slots for required attendees.
+        ArrayList<TimeRange> mandatorySlots = new ArrayList<>();
+        if(eventsTimeRangeList.size() > 0) {
+            mandatorySlots = findSlots(eventsTimeRangeList, request);
+        }
+
+        // Finds slots that are common to both requuired and optional attendees and returns them
+        // if there are any.
+        ArrayList<TimeRange> combinedSlots = new ArrayList<>();
+        if(optionalsSlots.size() > 0) {
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I don't think you need this if statement.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,119 @@
...
+  }
+
+  // Finds available TimeRange slots for a potential meeting given a meeting request.
+  private ArrayList<TimeRange> findSlots(ArrayList<TimeRange> eventsTimeRangeList, MeetingRequest request) {
+
+    ArrayList<TimeRange> timeSlots = new ArrayList<>();
+    Collections.sort(eventsTimeRangeList, TimeRange.ORDER_BY_START);
+
+    // Remove nested events.
+    for(int i = 0; i < eventsTimeRangeList.size()-1; i++) {
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This won't remove nested events completely if event with longer duration shows up later in array. Say:
[8:00, 9:00)
[8:00, 10:00)
Both will be retained.
Also we don't normally remove element in array while doing iteration.(Change array len during iteration)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,119 @@
...
+        int endTime   = eventsTimeRangeList.get(i+1).start();
+        long timeBetweenEvents = (long) (endTime - startTime);
+
+        // Checks if there is time overlaps, and if there is enough time to last the given meeting duration.
+        if( (startTime < endTime) && (timeBetweenEvents >= request.getDuration()) ) {
+            timeSlots.add( TimeRange.fromStartEnd(startTime, endTime, false) ); 
+        }        
+    }
+
+    // Add last TimeRange if the last TimeRange is not END_OF_DAY inclusive.
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:104]: If you have following ranges:
[8:00, END_OF_DAY)
[9:00, 11:00)
Will it return extra slot of [11:00, END_OF_DAY)?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,119 @@
...
+        ArrayList<TimeRange> mandatorySlots = new ArrayList<>();
+        if(eventsTimeRangeList.size() > 0) {
+            mandatorySlots = findSlots(eventsTimeRangeList, request);
+        }
+
+        // Finds slots that are common to both requuired and optional attendees and returns them
+        // if there are any.
+        ArrayList<TimeRange> combinedSlots = new ArrayList<>();
+        if(optionalsSlots.size() > 0) {
+            for(TimeRange optionalSlot : optionalsSlots) {
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: (Optional) This part is O(n^2). You can actually reduce it to O(n) by compare sorted TimeRange() -- Actually nlog(n) as sort itself is bounded by nlog(n)
Think about it but change is optional.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: using Collection here seems to generate an error, saying:
 ```incompatible types: java.util.Collection<com.google.sps.TimeRange> cannot be converted to java.util.ArrayList<com.google.sps.TimeRange> ```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: corrected

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: corrected

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yes; i have now deleted it.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: thanks for catching that. i am adding a functionality to tackle this case

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:104]: If the ranges you provided are timeranges where the attendees are busy elsewhere and are therefore not available to attend the meeting, it will return [START_OF_DAY, 8:00) only.

[11:00, END_OF_DAY) is not returned as an extra slot since not both of the attendees are available to attend at that time.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: added

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: i've implemented another approach to eliminate the need for this comparison here so as to avoid the o(n^2)

**Comment by kiprot** [comment]
Changes in new PR:
- changed approach to 
  -  first combine both required and optional attendees into one list and find slots for them
  - if there are no slots for a combination of both, i find slots for required attendees only, and return
  - if there are no slots for required attendees, i find if there are slots for optional attendees, and return that, or else an empty list is returned

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (octopus1000) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: As the error indicated, it failed because you are trying to assign array to collection.
By "using Collection" means you have to create collection, add/remove using collection operations as their implementation is different, not simply change the data type.

You can see the return value for query() function is Collection<TimeRange>, see how that is defined and operated.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +279,143 @@ public void notEnoughRoom() {
...
+      request.addOptionalAttendee(OPTIONAL_B);
+
+      Collection<TimeRange> actual = query.query(events, request);
+      Collection<TimeRange> expected = Arrays.asList();
+      Assert.assertEquals(expected, actual);
+  }
+
+  // testCase1 and testCase2 are tests to try the query for a potential edge case.
+  @Test
+  public void testCase1() {
```
Comment (octopus1000) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:382]: Give test a meaningful name.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: That's right. 
I am also using a .get() method somewhere in the code, which is in the ArrayList class, but not available in the Collection interface (which is the reason why I thought to use the Arraylist). 
Trying to implement a function that works like the get() method seems to be quite expensive since it might involve converting the Collection to an array, or using an Iterator every time.

[side-note question:] is it generally bad practice to return, say ArrayList/ Set/ List in a function that should return Collection, even though they implement the Collection interface?

Comment (octopus1000) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: Let's keep ArrayList here for now, as you need it for ordered list. The only drawback is you could possibly return duplicate slots. 

Actually there's something wrong with my previous comments(I wasn't aware collection is interface rather than concrete class). What I mean is a data type that will not contain duplicate elements, I think in Java, that's Set instead of Collection.

As to this error:
incompatible types: java.util.Collection<com.google.sps.TimeRange> cannot be converted to java.util.ArrayList<com.google.sps.TimeRange>
It happens because we are trying to cast Collection to ArrayList which is not feasible. You can cast class to interface but not the reverse way.

As to [side-note], as collection is interface, it's underlying class will always be a concrete class that implements the interface. So that's not a bad practice, that's the way it works. However you should set expectation for your return, do I want to return duplicate entry or not? If yes, it's totally fine to keep ArrayList, if not you might want to remove duplicates before return.

**Review by octopus1000 — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (kiprot) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: Yes, i should have used a set to avoid duplicates. I'll keep that in mind, and just merge for now.
